climada/engine/uncertainty/base.py

```diff
@@ -0,0 +1,728 @@
+            parameter names (all labels) from all unc_vars.
+            The default is pd.DataFrame().
+        metrics : dict(), optional
+            Dictionnary of the CLIMADA metrics for which the sensitivity
+            will be computed. For each sample, each metric must have a
+            definite value.
+            Keys are metric names (e.g. 'aai_agg', 'freq_curve') and 
+            values are pd.DataFrame with values for each parameter sample
+            (one row per sample).
+            The default is {}.
+        sensitivity: dict(), optional
+            Dictionnary of the sensitivity analysis for each uncertainty
+            parameter.
+            The default is {}.
+        """
+    
+        self.unc_vars = unc_vars if unc_vars else {}
+        self.sample = sample if sample is not None else pd.DataFrame(
+            columns = self.param_labels)
+        self.sampling_method = None
+        self.metrics = metrics if metrics is not None else {}
+        self.sensitivity = sensitivity if sensitivity is not None else {}
+        self.check()
+    
+    
+    def check(self):
+        """
+        Check if the data variables are consistent
+
+        Returns
+        -------
+        check: boolean
+            True if data is consistent.
+
+        """
+        check = True
+        check &= (self.param_labels == self.sample.columns.to_list())
+        if not check:
+            raise ValueError("Parameter names from unc_vars do not "
+                             "correspond to parameters names of sample")
+        for metric, df_distr in self.metrics.items():
+            if df_distr.empty:
+                continue
+            check &= (len(df_distr) == self.n_samples)
+            if not check:
+                raise ValueError(f"Metric f{metric} has less values than the "
+                             "number of samples {self.n_samples}")
+            
+            if df_distr.isnull().values.any():
+                LOGGER.warning("At least one metric evaluated to Nan for " +
+                    "one cominbation of uncertainty parameters containend " +
+                    "in sample. Note that the sensitivity analysis will " +
+                    "then return Nan. " +
+                    "See https://github.com/SALib/SALib/issues/237")
+        return check
+        
+    
+    @property
+    def n_samples(self):
+        """
+        The effective number of samples
+
+        Returns
+        -------
+        n_samples: int
+            effective number of samples
+
+        """
+
+        if isinstance(self.sample, pd.DataFrame):
+            return self.sample.shape[0]
+        return 0
+
+    @property
+    def param_labels(self):
+        """
+        Labels of all uncertainty parameters.
+
+        Returns
+        -------
+        param_labels: list of strings
+            Labels of all uncertainty parameters.
+
+        """
+        return list(self.distr_dict.keys())
```

Comment:
This is simply more flexible. If one can avoid unnecessary hard coding of variables that are actually a function of another variable, then one should not do it imho.

Reply:
Fair in general. This would mean the property is only assigned as an attribute to the parent class if the function which uses it, is actually called, right?
Two things that then speak against it: `param_labels` will always be used in the course of the uncertainty calculation, won't it? so it wouldn't unnecessarily "crowd" the `__init__` statement. Plus, assigning via decorators is extremely un-intuitive and makes it very very hard to directly see the structure / properties of a class for a first-time or not heavy user of this class. 

I think we've had the conversation about use of decorators earlier on in our coding convention meetings, with different opinions on that topic..

Reply:
>Fair in general. This would mean the property is only assigned as an attribute to the parent class if the function which uses it, is actually called, right?

That is not really correct. The attribute is always present, it just that the data is not saved in a separate memory slot, but retrieved from the already existing one (with maybe a computation in between).

One can of course define them in the `__init__`, but it is less flexible. Say you initiate your Object, then for some reason decide to add an uncertainty variable afterwards by modifying the `distr_dict`. For example, think of the case discussed with default values. You define the object, but then change `distr_dict = {'x':scipy.stats.norm(0,1)}` to `distri_dict = {'x':scipy.stats.norm(0,1), 'y':scipy.stats.norm(0,1)}`, then the code only works with the `@property`, otherwise note. I admit that this is not necessarily a common use case though.

Reply:
Right, i remember this from the pythonmorsel :D 
yeah well, alright then. I guess it has a certain advantage that may outweigh the readability downside, then.